editoast: refactor projects tests

Wadjetz opened the pull request.

hamz2a reviewed (approved):
LGTM, thanks !

leovalais commented on editoast/src/modelsv2/projects.rs:
```diff
@@ -118,84 +118,101 @@ impl Project {
+        let project_budget = Some(1000);
+
+        // Patch a project
+        let update_project = created_project
```
You could be using `.patch()` for this:

```rust
project.patch().name(...).budget(...).apply(conn).await
```

leovalais commented on editoast/src/modelsv2/projects.rs:
```diff
@@ -118,84 +118,101 @@ impl Project {
+            .expect("Project not found");
+
+        assert_eq!(&created_project, &project);
+
+        assert!(
+            Project::list(db_pool.get_ok().deref_mut(), SelectionSettings::new())
+                .await
+                .is_ok()
+        );
```
```suggestion
```

list is tested in another test below

leovalais commented on editoast/src/modelsv2/projects.rs:
```diff
@@ -118,84 +118,101 @@ impl Project {
 pub mod test {
     use rstest::rstest;
-    use std::sync::Arc;
+    use std::ops::DerefMut;
```
```suggestion
    use std::ops::DerefMut;
    use pretty_assertions::assert_eq;
```

It's a drop-in replacement for `assert_eq!`, but shows a colored diff on failure. It's already in the dependance tree so we might as well use it.

leovalais commented on editoast/src/views/projects.rs:
```diff
@@ -301,102 +302,166 @@ impl From<ProjectPatchForm> for Changeset<Project> {
+    use crate::modelsv2::prelude::*;
+    use crate::views::test_app::TestAppBuilder;
+
+    pub fn project_changeset() -> Changeset<Project> {
+        Project::changeset()
+            .name("test_project".to_owned())
+            .budget(Some(0))
+            .creation_date(Utc::now().naive_utc())
+            .last_modification(Utc::now().naive_utc())
+            .tags(Tags::default())
+    }
+
+    async fn create_project(pool: &DbConnectionPoolV2) -> Project {
+        project_changeset()
+            .create(pool.get_ok().deref_mut())
+            .await
+            .expect("Failed to create project")
```
This is mostly duplication of the functions in `modelsv2::project::test`. You could define a `modelsv2::fixtures` public module that could be reused in the views.

leovalais commented on editoast/src/modelsv2/projects.rs:
```diff
@@ -118,84 +118,101 @@ impl Project {
+            .tags(Tags::default())
+    }
+
+    async fn create_project(db_pool: &DbConnectionPoolV2, name: &str) -> Project {
```
```suggestion
    async fn create_project(conn: &mut DbConnection, name: &str) -> Project {
```

leovalais commented on editoast/src/views/projects.rs:
```diff
@@ -301,102 +302,166 @@ impl From<ProjectPatchForm> for Changeset<Project> {
+            .find(|p| p.project.id == created_project.id)
+            .unwrap();
+
+        assert_eq!(created_project.name, project_retreived.project.name);
```
```suggestion
        assert_eq!(created_project, project_retreived.project);
```

leovalais commented on editoast/src/views/projects.rs:
```diff
@@ -301,102 +302,166 @@ impl From<ProjectPatchForm> for Changeset<Project> {
+            .expect("Failed to retrieve project")
+            .expect("Project not found");
+
+        assert_eq!(project.name, updated_name);
```
```suggestion
        assert_eq!(project.name, updated_name);
        assert_eq!(project.budget, Some(updated_budget));
```

leovalais commented on editoast/src/views/pagination.rs:
```diff
@@ -47,7 +47,7 @@ const DEFAULT_PAGE_SIZE: u64 = 25;
 /// We named the data field `result` to cope with the old pagination schema which
 /// enforced this name. For new paginated responses, the field name is up to your imagination :)
-#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, ToSchema)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
```
```suggestion
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, ToSchema)]
#[cfg_attr(test, derive(Deserialize))]
```

same for other `Deserialize`

leovalais reviewed (approved):
Thanks for refactoring and improving these tests :)

I have a few comments but nothing really blocking (except for the `cfg_attr` one).

Wadjetz committed:
editoast: refactor projects tests

Update editoast/src/views/pagination.rs

Co-authored-by: Léo Valais <leo.valais97@gmail.com>
Update editoast/src/modelsv2/projects.rs

Co-authored-by: Léo Valais <leo.valais97@gmail.com>
Update editoast/src/modelsv2/projects.rs

leovalais commented on editoast/src/modelsv2/fixtures.rs:
```diff
@@ -0,0 +1,25 @@
+#[cfg(test)]
+pub mod tests {
```
You don't need to make a test submodule, just `#[cfg(test)]` its inclusion in the source tree

leovalais commented on editoast/src/modelsv2/fixtures.rs:
```diff
@@ -0,0 +1,25 @@
+    use crate::modelsv2::Project;
+    use crate::modelsv2::Tags;
+
+    pub fn test_project_changeset(name: &str) -> Changeset<Project> {
```
Completely subjective comment + nitpick: I'd drop the `test` prefix as the function module already suggests we're in a test context (we have namespaces for a reason ^^)

leovalais commented on editoast/src/modelsv2/mod.rs:
```diff
@@ -1,6 +1,7 @@
 pub mod database;
 pub mod documents;
 pub mod electrical_profiles;
+pub mod fixtures;
```
```suggestion
#[cfg(test)] pub mod fixtures;
```

leovalais commented on editoast/src/modelsv2/projects.rs:
```diff
@@ -117,85 +117,86 @@ impl Project {
             .clone()
-            .into_changeset()
-            .name(project.model.name.clone() + "_bis");
+            .patch()
```
you don't need to clone() it (am I missing something? 🤔)

Wadjetz replied:
No, it wasn't necessary to clone. 👍

leovalais commented on editoast/src/views/projects.rs:
```diff
@@ -120,7 +119,7 @@ async fn check_image_content(db_pool: Data<DbConnectionPool>, document_key: i64)
 }
 
-#[derive(Debug, Clone, Serialize, ToSchema)]
+#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
```
cfg_attr

leovalais reviewed (commented):
A few more 😄
